File: app/src/main/java/Day07.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Day07 {
    public static String compress(String s) {
       return usingStreams(s);
    }

    private static String usingLoops(String s) {
        if(s.isEmpty()) {
            return s;
        }

        StringBuilder compressedS = new StringBuilder();

        char last = s.charAt(0);
        int count = 1;

        for(int i = 1; i < s.length() ; i++) {
            if(last == s.charAt(i)) {
                count++;
            } else {
                compressedS.append(last).append(count);
                last = s.charAt(i);
                count = 1;
            }
        }

        //capture the last part of the string
        compressedS.append(last).append(count);

        return compressedS.toString();
    }

    // Stream-based implementation of run-length encoding using a stateful collector
    // classic loop approach better choice due to complexity of stateful element
    private static String usingStreams(String s) {
        if (s.isEmpty()) return s;

        return s.chars()
                .mapToObj(c -> (char) c)
                .collect(
                        ArrayList<int[]>::new,
                        (list, c) -> {
                            if (list.isEmpty() || list.get(list.size() - 1)[0] != c)
                                list.add(new int[]{c, 1});
                            else
                                list.get(list.size() - 1)[1]++;
                        },
                        List::addAll
                ).stream()
                .map(a -> "" + (char) a[0] + a[1])
                .collect(Collectors.joining());
    }
}

File: app/src/main/kotlin/Day07.kt
fun compress(s: String) : String = usingLoop(s)

private fun usingLoop(s: String): String {
    if (s.isEmpty()) return s

    val compressedS = StringBuilder()
    var count = 1

    for (i in 1 until s.length) {
        if (s[i] == s[i - 1]) {
            count++
        } else {
            compressedS.append(s[i - 1]).append(count)
            count = 1
        }
    }

    compressedS.append(s.last()).append(count)
    return compressedS.toString()
}

//uses fold to accumulate and Pair to keep track of last character/count
private fun usingExtensions(s: String): String =
    s.fold(mutableListOf<Pair<Char, Int>>()) { acc, c ->
        if (acc.isEmpty() || acc.last().first != c)
            acc.add(c to 1)
        else
            acc[acc.lastIndex] = c to acc.last().second + 1
        acc
    }.joinToString("") { "${it.first}${it.second}" }